Ensure `@config` is injected in common ancestor sheet

RobinMalfait opened the pull request:
This PR fixes an issue where an `@config` was injected in a strange location if you have multiple CSS files with Tailwind directives. 

Let's say you have this setup:
```css
/* ./src/index.css */
@import "./tailwind-setup.css";

/* ./src/tailwind-setup.css */
@import "./base.css";
@import "./components.css";
@import "./utilities.css";

/* ./src/base.css */
@tailwind base;

/* ./src/components.css */
@tailwind components;

/* ./src/utilities.css */
@tailwind utilities;
```

In this case, `base.css`, `components.css`, and `utilities.css` are all considered Tailwind roots because they contain Tailwind directives or imports.

Since there are multiple roots, the nearest common ancestor should become the tailwind root (where `@config` is injected). In this case, the nearest common ancestor is `tailwind-setup.css` (not `index.css` because that's further away).

Before this change, we find the common ancestor between `base.css` and `components.css` which would be `index.css` instead of `tailwind-setup.css`.

In a next iteration, we compare `index.css` with `utilities.css` and find that there is no common ancestor (because the `index.css` file has no parents). This resulted in the `@config` being injected in `index.css` and in `utilities.css`.

Continuing with the rest of the migrations, we migrate the `index.css`'s `@config` away, but we didn't migrate the `@config` from `utilities.css`.

With this PR, we don't even have the `@config` in the `utilities.css` file anymore.

Test plan
---

1. Added an integration test with a non-migrateable config file to ensure that the `@config` is injected in the correct file.
2. Added an integration test with a migrateable config file to ensure that the CSS config is injected in the correct file. h/t @philipp-spiess 
3. Ran the upgrade on the https://commit.tailwindui.com project and ensured that 
  1. The `@config` does not exist in the `utilities.css` file (this was the first bug we solved)
  2. The `@config` is replaced in the `tailwind.css` file correctly.

<img width="592" alt="image" src="https://github.com/user-attachments/assets/02e3f6ea-a85d-46c2-ac93-09f34ac4a4b8">

<img width="573" alt="image" src="https://github.com/user-attachments/assets/e372eb5f-5732-4052-ab39-096ba7970ff6">

RobinMalfait committed:
break once we found common parent

Once we found a common parent between two sheets, we can stop finding
the parent and continue to the next sheets.

RobinMalfait committed:
add failing test

RobinMalfait committed:
update changelog

philipp-spiess commented on integrations/upgrade/index.test.ts:
```diff
@@ -1467,6 +1467,139 @@ test(
+      'tailwind.config.ts': js`
+        export default {
+          content: ['./src/**/*.{html,js}'],
+          plugins: [
+            () => {
+              // custom stuff which is too complicated to migrate to CSS
+            },
+          ],
```
```suggestion
          theme: {
            extend: {
              colors: {
                'my-red': 'red',
              },
            },
          },
```

When we update the config to something that we _can_ convert to, we noticed that the `@theme` now ends up in `src/index.css` and that the `@config` is still inside `src/tailwind-setup.css`.

RobinMalfait replied:
Added a failing test for this: https://github.com/tailwindlabs/tailwindcss/pull/14989/commits/a8a66a4aa62ad7d79bcc9a23ef9e104b47498f0e

Then fixed it: https://github.com/tailwindlabs/tailwindcss/pull/14989/commits/dc5cd8db12545c89852a2dae853d9193f537210f

Bonus points: basically deleted most of the code in the `migrate-config` file

RobinMalfait committed:
add failing test

RobinMalfait committed:
always migrate the js config in the Tailwind Root file

We already computed the correct Tailwind root file (and linked the JS
config to it). This means that we can remove the logic in the
migrate-config migration to find the correct location.

RobinMalfait committed:
run prettier

RobinMalfait committed:
Merge branch 'next' into fix/incorrect-config-link

philipp-spiess commented on packages/tailwindcss/src/compat/plugin-api.test.ts:
```diff
@@ -2973,8 +2973,7 @@ describe('matchUtilities()', () => {
       return compiled.build(candidates)
     }
 
-    expect(optimizeCss(await run(['@w-1','hover:@w-1'])).trim())
```
Oh how did I smuggle that in huh

philipp-spiess commented on integrations/upgrade/index.test.ts:
```diff
@@ -1467,6 +1467,276 @@ test(
+      }
+
+      --- ./src/components.css ---
+      @import './typography.css';
```
Just for completeness: This is still wrong, we expect this to be moved into a layer. This is planned for a follow-up PR.

philipp-spiess commented on packages/@tailwindcss-upgrade/src/codemods/migrate-config.ts:
```diff
@@ -82,41 +84,6 @@ export function migrateConfig(
     root.append(cssConfig.nodes)
   }
 
-  function migrate(root: Root) {
```
very nice cleanup
